primitives/parachain-inherent/src/client_side.rs

```diff
@@ -99,6 +108,26 @@ async fn collect_relay_storage_proof(
 	relevant_keys.push(relay_well_known_keys::CURRENT_SLOT.to_vec());
 	relevant_keys.push(relay_well_known_keys::ACTIVE_CONFIG.to_vec());
 	relevant_keys.push(relay_well_known_keys::dmq_mqc_head(para_id));
+
+	tracing::debug!(
+		target: LOG_TARGET,
+		?dmp_message_window,
+		"seting dmq_mqc_head_for_message relevant keys",
+	);
+
+	// Collect proof for a subset of messages (`downward_messages_count`).
+	if let Some(first_message) = dmp_message_window.first() {
+		let first_message = first_message.message_idx;
+		let last_message =
+			first_message.wrapping_add(downward_messages_count.into()).wrapping_dec();
+
+		for idx in first_message.into()..=last_message.into() {
```

Comment:
I believe the fix was not correct, let me try to elaborate:
1. First, not sure why here `wrapping_dec` is used to calculate `last_message` and not `wrapping_add`.
2. If with add, wouldn't it be `last_message + 1` and not actually `last_message`?
3. You're converting the `WrappingIndex` to u64 here IIUC, but note that 3..=2 is an empty loop. What you want here is either a while loop or implement a different type than `WrappingIndex` that can iterate as expected on wrap.

Reply:
1. because adding the amount of messages(N) to the first message index would return the index of the N+1 message.
2. I don't really know what you mean.
3. Yeah, this is wrong, I'll do it similarly how it works in DMP pallet. I don't think adding extra type is needed.

Reply:
Sorry about 1 or 2, I misread the code to assume it's last = first - count

Reply:
Regarding 1,2 - gonna change it a little bit to be more readable.

Reply:
@ordian please take another look, this should be fine now.

Reply:
Looks good now :+1: